### DISCUSSION TITLE
Feature Request: User Feedback after every chat message in the chat card

### DISCUSSION THREAD

**Discussion by shivam5992** [h2oai/wave#2233] [Ideas]
Scenario - In the Chat Card that may be connected to an LLM, Users need a functionality where they can vote (thumbs up/down, or select an emoji) about the quality of the response, they also want to open a text field to provide text feedback. 

--> 

<img width="602" alt="image" src="https://github.com/h2oai/wave/assets/4301965/50a7cf8b-a2b5-4bca-9a1a-51ed9ec0c789">

**Comment by mturoci** [comment]
https://github.com/h2oai/wave/pull/2178

**Reply by shivam5992** [reply]
Is it merged already? 
Is there ability ot add text feedback also?

**Reply by mturoci** [reply]
Yes, merged. Text feedback needs to be handled by app developer. Wave notifies you about feedback event to which you can respond with dialog that includes textbox etc.

**Comment by mtanco** [comment]
@mturoci any best practices about using this feature while streaming? or when using system prompts?

<img width="1025" alt="image" src="https://github.com/h2oai/wave/assets/4793561/2528381d-1c9f-4c9c-b961-0c4f44e6340d">

<img width="1010" alt="image" src="https://github.com/h2oai/wave/assets/4793561/7f27fe69-6e22-4cce-b1d0-4245ab62433d">

**Reply by mturoci** [reply]
> using this feature while streaming? or when using system prompts?

Do you face any problems in particular? The only best practice I can recommend atm is to handle the feedback event the way you want (store in DB, send to API etc). Should be independent of streaming/system prompts but maybe I am missing something?

**Comment by mtanco** [comment]
It’s really weird UX no? Here’s a message from someone introducing
themselves and I can downvote it, here’s a message actively streaming and I
have rating buttons, it’s unusually user behavior.

Michelle Gabriele Tanco

Product Management

***@***.*** | 1-(253)-293-8551

<https://twitter.com/MichelleTanco> <https://www.linkedin.com/mtanco>

Democratize Generative AI <https://www.h2o.ai/>

Try h2oGPT for free! <https://gpt.h2o.ai/>

**Reply by mturoci** [reply]
> Here’s a message from someone introducing themselves and I can downvote it

Agree, we should allow feedback for responses only.

> here’s a message actively streaming and I have rating buttons

I don't see a UX problem with this one. You can usually tell after a few sentences whether the answer is good or not, no need to wait  until completion. Can be changed if you feel strongly about it though.
